src/hmem_ze.c

```diff
@@ -594,6 +594,44 @@ static int ze_hmem_find_copy_only_engine(int device_num, int *ordinal)
 	return ze_ret;
 }
 
+static int ze_hmem_cleanup_internal(int fini_workaround)
+{
+	int i, ret = FI_SUCCESS;
+
+	for (i = 0; i < num_devices; i++) {
+		/*
+		 * zeCommandQueueDestroy() may segfault when called from
+		 * "destructor" if any hmem-capable DL provider is enabled
+		 * due to premature unloading of GPU specific L0 library.
+		 * Before a permanent fix is available, bypass this call
+		 * and let zeContextDestroy() handle the resource cleanup.
+		 */
+		if (!fini_workaround) {
+			if (cmd_queue[i] &&
+			    ofi_zeCommandQueueDestroy(cmd_queue[i])) {
+				FI_WARN(&core_prov, FI_LOG_CORE,
+					"Failed to destroy ZE cmd_queue\n");
+				ret = -FI_EINVAL;
+			}
+		}
+		if (dev_fds[i] != -1) {
+			close(dev_fds[i]);
+			dev_fds[i] = -1;
+		}
+	}
+
+	if (ofi_zeContextDestroy(context))
+		ret = -FI_EINVAL;
+
+	ze_hmem_dl_cleanup();
+	return ret;
+}
```

Comment:
@aingerson Can this cleanup happen earlier in the code flow?  For example, when the fabric or last ep is closed?

Reply:
Nothing prevents application from opening multiple fabric objects.

Reply:
I meant the last fabric is closed.

Reply:
The application can open a new fabric after closing an old one.

Reply:
Yes, this would require moving the initialization code so that the GPU resources are opened/closed in conjunction with a specific OFI resource (fabric, domain, ep...)